src/OpenLoco/src/Vehicles/Vehicle.cpp

```diff
@@ -115,14 +130,199 @@ namespace OpenLoco::Vehicles
+            while (component.remainingDistance >= 0x368A)
+            {
+                hasMoved = true;
+                auto newSubPosition = component.subPosition + 1U;
+                const auto subPositionDataSize = World::TrackData::getTrackSubPositon(component.trackAndDirection.track._data).size();
+                // This means we have moved forward by a track piece
+                if (newSubPosition >= subPositionDataSize)
+                {
+                    if (!updateTrackMotionNewTrackPiece(component))
+                    {
+                        returnValue = component.remainingDistance - 0x3689;
+                        component.remainingDistance = 0x3689;
+                        _vehicleUpdate_var_1136114 |= (1U << 0);
+                        break;
+                    }
+                    else
+                    {
+                        newSubPosition = 0;
+                    }
+                }
+                // 0x004B1761
+                component.subPosition = newSubPosition;
+                const auto moveData = World::TrackData::getTrackSubPositon(component.trackAndDirection.track._data)[newSubPosition];
+                const auto nextNewPosition = moveData.loc + World::Pos3(component.tileX, component.tileY, component.tileBaseZ * World::kSmallZStep);
+                component.remainingDistance -= movementNibbleToDistance[getMovementNibble(intermediatePosition, nextNewPosition)];
+                intermediatePosition = nextNewPosition;
+                component.spriteYaw = moveData.yaw;
+                component.spritePitch = moveData.pitch;
+                if (component.getSubType() == VehicleEntityType::bogie)
+                {
+                    // collision checks
+                    auto collideResult = checkForCollisions(component, intermediatePosition);
+                    if (collideResult.has_value())
+                    {
+                        _vehicleUpdate_var_1136114 |= (1U << 2);
+                        _vehicleUpdate_collisionCarComponent = collideResult.value();
+                    }
+                }
+            }
+            if (hasMoved)
+            {
+                Ui::ViewportManager::invalidate(&component, ZoomLevel::eighth);
+                component.moveTo(intermediatePosition);
+                Ui::ViewportManager::invalidate(&component, ZoomLevel::eighth);
+            }
+            return returnValue;
+        }
+        else
+        {
+            assert(false);
+            return 0;
+        }
+    }
+
     // 0x004B15FF
     uint32_t VehicleBase::updateTrackMotion(uint32_t unk1)
     {
-        registers regs;
-        regs.eax = unk1;
-        regs.esi = X86Pointer(this);
-        call(0x004B15FF, regs);
-        return regs.eax;
+        switch (getSubType())
+        {
+            case VehicleEntityType::head:
+                return updateTrackMotionT(*reinterpret_cast<VehicleHead*>(this), unk1);
```

Comment:
I think we have a function `as` for this, having a visitor pattern for this would be a pretty good thing which skips the entire checking for the type then, I would propose to just use `as` for now and then we should look into adding `visitSubType` or something to make patterns like this easier.

Reply:
I decided to do this differently so no longer using templates.